### PR TITLE
fix: [CDS-51574]: handled batched delete of approvalInstances when related nodes are deleted

### PR DESCRIPTION
**Pull request by naman-goenka**
## Describe your changes
- batched delete of approval Instances when a associated nodeExecution is being deleted.
- MAX_BATCH_SIZE = 500
- Testing: junits, local testing by deleting org and project associated approvalInstances also deleted.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43210)

### CODE REVIEW

**Review by archit-harness — CHANGES_REQUESTED** [review]
Done
> Reviewed 2 of 8 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43210)* status: 8 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/repositories/ApprovalInstanceCustomRepository.java` line 23 at r2](https://reviewable.io/reviews/harness/harness-core/43210#-NNCCyBo3_kcGtXhhq3L:-NNCCyBp4LKjDXqAldfb:b8wjm3u) ([raw file](https://github.com/harness/harness-core/blob/dcb3676fa3764dc5f72c8f2f32f1fc9ebd98e2f9/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/repositories/ApprovalInstanceCustomRepository.java#L23)):*
> ```Java
>   ApprovalInstance updateFirst(Query query, Update update);
>   UpdateResult updateMulti(Query query, Update update);
>   long deleteMulti(@NotNull Query query);
> ```

you can write delete using spring repository rather than this
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java` line 250 at r2](https://reviewable.io/reviews/harness/harness-core/43210#-NNCD5o764hPYWz1Dv72:-NNCD5o764hPYWz1Dv73:boj8hdr) ([raw file](https://github.com/harness/harness-core/blob/dcb3676fa3764dc5f72c8f2f32f1fc9ebd98e2f9/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java#L250)):*
> ```Java
>       // uses nodeExecutionId_1 idx
>       long deletedCount = approvalInstanceRepository.deleteMulti(
>           new Query(Criteria.where(ApprovalInstanceKeys.nodeExecutionId).in(nodeExecutionIds)));
> ```

use spring repository



_Code quote:_
```Java
      long deletedCount = approvalInstanceRepository.deleteMulti(
          new Query(Criteria.where(ApprovalInstanceKeys.nodeExecutionId).in(nodeExecutionIds)));
```

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java [hunk 1]
```diff
@@ -200,6 +205,54 @@ public HarnessApprovalInstance addHarnessApprovalActivity(@NotNull String approv
...
+  @Override
+  public boolean isNodeExecutionOfApprovalStepType(NodeExecution nodeExecution) {
+    if (isNull(nodeExecution) || isNull(nodeExecution.getStepType())) {
+      return false;
+    }
+    if (!StepCategory.STEP.equals(nodeExecution.getStepType().getStepCategory())) {
+      return false;
+    }
+    String stepType = nodeExecution.getStepType().getType();
+    return StepSpecTypeConstants.HARNESS_APPROVAL.equals(stepType)
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java:217]: Create a static set on top of this class & check contains.

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java [hunk 2]
```diff
@@ -200,6 +205,54 @@ public HarnessApprovalInstance addHarnessApprovalActivity(@NotNull String approv
...
+   * in a batch operation
+   * @param nodeExecutionIds
+   */
+  @Override
+  public void deleteByNodeExecutionIds(@NotNull Set<String> nodeExecutionIds) {
+    if (isEmpty(nodeExecutionIds)) {
+      return;
+    }
+    Set<String> batchNodeExecutionIds = new HashSet<>();
+    for (String batchNodeExecutionId : nodeExecutionIds) {
```
Comment (puthrayaharness) [inline pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java:234]: Consider using https://stackoverflow.com/questions/12026885/is-there-a-common-java-utility-to-break-a-list-into-batches
suggestions to create the batches. Then loop over each batch and call function

**Review by puthrayaharness — CHANGES_REQUESTED** [review]

File: pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java (hunk 1 above)
Comment (naman-goenka) [reply pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java:217]: done

**Comment by naman-goenka** [comment]
> > Reviewed 2 of 8 files at r1, all commit messages.
> > _[Reviewable](https://reviewable.io/reviews/harness/harness-core/43210)_ status: 8 file(s) to review. 1 change requests. 0 of 2 approvals obtained.
> 
> _[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/repositories/ApprovalInstanceCustomRepository.java` line 23 at r2](https://reviewable.io/reviews/harness/harness-core/43210#-NNCCyBo3_kcGtXhhq3L:-NNCCyBp4LKjDXqAldfb:b8wjm3u) ([raw file](https://github.com/harness/harness-core/blob/dcb3676fa3764dc5f72c8f2f32f1fc9ebd98e2f9/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/repositories/ApprovalInstanceCustomRepository.java#L23)):_
> 
> > ```java
> >   ApprovalInstance updateFirst(Query query, Update update);
> >   UpdateResult updateMulti(Query query, Update update);
> >   long deleteMulti(@NotNull Query query);
> > ```
> 
> you can write delete using spring repository rather than this
> 
> _[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java` line 250 at r2](https://reviewable.io/reviews/harness/harness-core/43210#-NNCD5o764hPYWz1Dv72:-NNCD5o764hPYWz1Dv73:boj8hdr) ([raw file](https://github.com/harness/harness-core/blob/dcb3676fa3764dc5f72c8f2f32f1fc9ebd98e2f9/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java#L250)):_
> 
> > ```java
> >       // uses nodeExecutionId_1 idx
> >       long deletedCount = approvalInstanceRepository.deleteMulti(
> >           new Query(Criteria.where(ApprovalInstanceKeys.nodeExecutionId).in(nodeExecutionIds)));
> > ```
> 
> use spring repository
> 
> _Code quote:_
> 
> ```java
>       long deletedCount = approvalInstanceRepository.deleteMulti(
>           new Query(Criteria.where(ApprovalInstanceKeys.nodeExecutionId).in(nodeExecutionIds)));
> ```

Done

**Review by puthrayaharness — APPROVED** [review]

**Review by puthrayaharness — APPROVED** [review]
> Reviewed 2 of 8 files at r1, 1 of 1 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43210)* status: 8 file(s) to review. 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — APPROVED** [review]
> Reviewed 1 of 8 files at r1, 1 of 1 files at r2, 6 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43210)* status: :shipit: complete! ✓

___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java` line 250 at r2](https://reviewable.io/reviews/harness/harness-core/43210#-NNCD5o764hPYWz1Dv72:-NNFgHeE1c8QxBd9oujA:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/dcb3676fa3764dc5f72c8f2f32f1fc9ebd98e2f9/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/steps/approval/step/ApprovalInstanceServiceImpl.java#L250)):*
<details><summary><i>Previously, naman-goenka (Naman Goenka) wrote…</i></summary><blockquote>

Done
</blockquote></details>

done
___
*[`pipeline-service/modules/orchestration-steps/src/main/java/io/harness/repositories/ApprovalInstanceCustomRepository.java` line 23 at r2](https://reviewable.io/reviews/harness/harness-core/43210#-NNCCyBo3_kcGtXhhq3L:-NNFgMgxB81zw2pxIBAw:bb7f2mf) ([raw file](https://github.com/harness/harness-core/blob/dcb3676fa3764dc5f72c8f2f32f1fc9ebd98e2f9/pipeline-service/modules/orchestration-steps/src/main/java/io/harness/repositories/ApprovalInstanceCustomRepository.java#L23)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

you can write delete using spring repository rather than this
</blockquote></details>

done

**Pull request merged**